Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,165 @@
 
 import NIOConcurrencyHelpers
 
+private struct SocketChannelLifecycleManager {
+    // MARK: Types
+    private enum State {
+        case fresh
+        case registered
+        case activated
+        case closed
+    }
+
+    private enum Event {
+        case activate
+        case register
+        case close
+    }
+
+    // MARK: properties
+    // this is queried from the Channel, ie. must be thread-safe
+    internal let isActiveAtomic: Atomic<Bool>
+    // these are only to be accessed on the EventLoop
+    internal let channelPipeline: ChannelPipeline
+    private let eventLoop: EventLoop
+    private var currentState: State = .fresh {
+        didSet {
+            assert(self.eventLoop.inEventLoop)
+            switch (oldValue, self.currentState) {
+            case (_, .activated):
+                self.isActiveAtomic.store(true)
+            case (.activated, _):
+                self.isActiveAtomic.store(false)
+            default:
+                ()
+            }
+        }
+    }
+
+    // MARK: API
+    internal init(eventLoop: EventLoop,
+                  channelPipeline: ChannelPipeline,
+                  isActiveAtomic: Atomic<Bool>) {
+        self.isActiveAtomic = isActiveAtomic
+        self.channelPipeline = channelPipeline
+        self.eventLoop = eventLoop
+    }
+
+    // this is called from Channel's deinit, so don't assert we're on the EventLoop!
+    internal var canBeDestroyed: Bool {
+        if case .closed = self.currentState {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    @inline(__always)
+    internal mutating func register(promise: EventLoopPromise<()>?) -> (() -> Void) {
+        return self.moveState(event: .register, promise: promise)
+    }
+
+    @inline(__always)
+    internal mutating func close(promise: EventLoopPromise<()>?) -> (() -> Void) {
+        return self.moveState(event: .close, promise: promise)
+    }
+
+    @inline(__always)
+    internal mutating func activate(promise: EventLoopPromise<()>?) -> (() -> Void) {
+        return self.moveState(event: .activate, promise: promise)
+    }
+
+    // MARK: private API
+    @inline(__always) // so we're not actually returning a closure
```

Comment:
So just to clarify this is to safe the heap allocation ? If so can you just add this to the comment ?